File: 2962_Count_Subarrays_Where_Max_Element_Appears_at_Least_K_Times/main.py
class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        left, right = 0, 0
        l = len(nums)
        cnt = 0
        ans = 0
        maxNum = max(nums)

        total = nums.count(maxNum)
        if total < k:
            return 0

        while left <= right:
            while cnt >= k:
                if nums[left] == maxNum:
                    cnt -= 1
                left += 1
                if cnt >= k:
                    ans += l - right + 1

            while cnt < k and right < l:
                if nums[right] == maxNum:
                    cnt += 1
                right += 1

            if cnt < k:
                return ans

            ans += l - right + 1

        return ans

File: 2962_Count_Subarrays_Where_Max_Element_Appears_at_Least_K_Times/main.go
func countSubarrays(nums []int, k int) int64 {
	left, right := 0, 0
	l := len(nums)
	cnt := 0
	ans := 0
	maxNum := 0
	for _, v := range nums {
		if v > maxNum {
			maxNum = v
		}
	}

	for left <= right {
		for cnt >= k {
			if nums[left] == maxNum {
				cnt--
			}
			left++
			if cnt >= k {
				ans += l - right + 1
			}
		}
		for right < l && cnt < k {
			if nums[right] == maxNum {
				cnt++
			}
			right++
		}

		if cnt < k {
			return int64(ans)
		}
		ans += l - right + 1
	}

	return int64(ans)
}
